doc/developer/platform/reconciliation.md

```diff
@@ -0,0 +1,64 @@
+# Reconciliation of the ADAPTER/COMPUTE communication
+
+The current proposal for Materialize's platform includes the separation of the ADAPTER and COMPUTE components into distinct processes.
+Each represents its own failure domain and hence we need to design the communication protocols around failures and active replication (AR).
+In general, we need to define how commands and responses can be reconciled.
+
+## Command reconciliation
+
+Commands represent instructions the COMPUTE controller sends to COMPUTE instances.
+* The COMPUTE controller may restart independently of a COMPUTE instance.
+* Reconciling commands is the mechanism to match a COMPUTE instance's state to what the COMPUTE controller expects after startup.
+* On startup, the COMPUTE controller provides the current configuration to the COMPUTE instances.
+* COMPUTE instances can match the configuration to their own state and only apply changes.
+
+For this purpose, we interpret the commands COMPUTE controller provides to COMPUTE as an append-only log.
+Upon COMPUTE controller restart, the new configuration is/should be a suffix of the log of all previously received commands.
+
+The log can be compacted by collapsing commands that semantically supersede previous commands.
+
+Currently, the `Command` protocol does not express enough information to provide full command reconciliation.
+* Materialize initiates tables with ephemeral information.
+  For example, `show databases` reads from a table that is unconditionally initialized with `materialize`.
+  The reconciler has no good mechanism to distinguish initialization updates from others.
+  * *Current solution:* The coordinator truncates all ephemeral tables on startup using a new instruction.
```

Comment:
I'm not sure I understand how the non-truncated tables work. For example, if the controller goes down, we won't know which commands made it through to the storage layer, and on restart .. the contents of these tables will be in an unclear state. Are the "non-ephemeral" tables .. just ones with random stuff in them?